nemo/collections/asr/modules/graph_decoder.py

```diff
@@ -0,0 +1,167 @@
+        return_type: str = "1best",
+        return_ilabels: bool = True,
+        output_aligned: bool = True,
+        split_batch_size: int = 0,
+        graph_module_cfg: Optional[DictConfig] = None,
+    ):
+        self._blank = num_classes
+        self.return_ilabels = return_ilabels
+        self.output_aligned = output_aligned
+        self.split_batch_size = split_batch_size
+        self.dec_type = dec_type
+
+        if return_type == "1best":
+            self.return_lattices = False
+        elif return_type == "lattice":
+            self.return_lattices = True
+        elif return_type == "nbest":
+            raise NotImplementedError(f"return_type {return_type} is not supported at the moment")
+        else:
+            raise ValueError(f"Unsupported return_type: {return_type}")
+
+        # we assume that self._blank + 1 == num_classes
+        if backend == "k2":
+            # use k2 import guard
+            from nemo.core.utils.k2_utils import k2_import_guard
+
+            k2_import_guard()
+
+            if self.dec_type == "topo":
+                from nemo.collections.asr.parts.k2.graph_decoders import BaseDecoder as Decoder
+            elif self.dec_type == "tokenlm":
+                from nemo.collections.asr.parts.k2.graph_decoders import TokenLMDecoder as Decoder
+            elif self.dec_type == "looseali":
+                raise NotImplementedError()
+            elif self.dec_type == "tlg":
+                raise NotImplementedError(f"dec_type {dec_type} is not supported at the moment")
+            else:
+                raise ValueError(f"Unsupported dec_type: {dec_type}")
+
+            self._decoder = Decoder(num_classes=self._blank + 1, blank=self._blank, cfg=graph_module_cfg)
+        elif backend == "gtn":
+            raise NotImplementedError("gtn-backed decoding is not implemented")
+
+        super().__init__()
+
+    def update_graph(self, graph):
+        self._decoder.update_graph(graph)
+
+    def _forward_impl(self, log_probs, log_probs_length, targets=None, target_length=None):
+        if targets is None and target_length is not None or targets is not None and target_length is None:
+            raise RuntimeError(
+                f"Both targets and target_length have to be None or not None: {targets}, {target_length}"
+            )
+        # do not use self.return_lattices for now
+        if targets is None:
+            align = False
+            decode_func = lambda a, b: self._decoder.decode(
+                a, b, return_lattices=False, return_ilabels=self.return_ilabels, output_aligned=self.output_aligned
+            )
+        else:
+            align = True
+            decode_func = lambda a, b, c, d: self._decoder.align(
+                a, b, c, d, return_lattices=False, return_ilabels=False, output_aligned=True
+            )
+        batch_size = log_probs.shape[0]
+        if self.split_batch_size > 0 and self.split_batch_size < batch_size:
+            predictions = []
```

Comment:
Im just wondering (disregard if it makes no sense) - can we extract this subset of code out into a private function ? This is fine too, im wondering if there will ever be a future where a class needs to change only this part. Maybe its never going to happen.

Reply:
We can, but I've left it as is. I don't think this part will ever change.